openpype/plugins/publish/integrate_new.py

```diff
@@ -110,801 +156,494 @@ class IntegrateAssetNew(pyblish.api.InstancePlugin):
     exclude_families = ["clip"]
     db_representation_context_keys = [
         "project", "asset", "task", "subset", "version", "representation",
-        "family", "hierarchy", "task", "username"
+        "family", "hierarchy", "task", "username", "frame"
```

Comment:
The `frame` really should not be here.

Reply:
Looking at the code changes the behavior should be the same as before. `frame` is only added if a valid value is present for frame and if used for the sequence - this is because a single file pops out the frame data anyway.

Or what am I missing? Or why must this not be here?

Reply:
It should be added only if is used, not if is available.

Reply:
> It should be added only if is used, not if is available.

Could you explain by what it should be used? Is that used for destination template? Or for something else?

Currently they are only added when like this like before. This means these keys are added if in template data - not necessarily if used by the template. So you are correct - that sounds wrong if they only should be added if used. Thanks!
```
        for key in self.db_representation_context_keys:
            value = template_data.get(key)
            if not value:
                continue
            repre_context[key] = template_data[key]
```
_Which is the same logic as before [here](https://github.com/pypeclub/OpenPype/blob/develop/openpype/plugins/publish/integrate_new.py#L562-L566)._

The `frame` key would never be added if not used. It's always used in the Sequence/Collection, but never for a single file... and in that case it's popped. But I think you might be referring to that `frame` data would come [from the "used_values" here](https://github.com/pypeclub/OpenPype/blob/develop/openpype/plugins/publish/integrate_new.py#L441-L442). The interesting bit being, that if we only want to store if used - why do we need to be explicit for the others with the `db_representation_context_keys`? Do we always **require** `hierarchy`, `task`, etc even when NOT used?

Reply:
It `template_data` may contain `"frame"` but template may not have `{frame}` in it.

Reply:
> The rest of data are to be able get context from representation itself when something goes wrong (e.g. see in what context it was published against in what context is now).

Check. Totally agree - unused data should be ignored.
So the other keys from `db_representation_context_keys` sounds like are absolutely required instead of optional? 

Or are they "optional" for that we only add them if they were set we put them in the context even if unused by template? That is the current and old behavior. That does still allow the cases that we can't rely on those values being present.

---

Anyway, sounds like we could get by with this:

```
# Always add to representation context even if not used by template
self.representation_context_keys = [
    "project", "asset", "task", "subset", "version", "representation",
    "family", "hierarchy", "username"
]

# Only add to representation context if used by template for
# destination path formatting
self.representation_template_used_context_keys = ["frame", "udim"]
```

---

After all this time only now I see that `"task"` appears twice in `db_representation_context_keys` (also before the refactor) which seems redundant - will clean that up too.

Reply:
Removed `"frame"` from `db_representation_context_keys` with https://github.com/pypeclub/OpenPype/pull/2898/commits/3ec9684239b7afc326cad7e184a7c6ed4e7a6058

The `repre_context` always contains all used values - so that would still include "frame" only if it's used in destination template. Is this sufficient now?